fix(jest): support more config file formats

nicojs opened the pull request:
Support all config file formats for jest using jest's native config reading (including `jest.config.ts`). See https://jestjs.io/docs/configuration for a list of all available config formats.

Stryker only uses jest's native functionality if it finds `jest@>=29.2`, otherwise, it falls back on its custom config file reading algorithm. 

This also adds support for exporting an `async` function from your jest config file even for `jest@<29.2`, because of added functionality in the custom config loading logic.

See https://github.com/facebook/jest/pull/13356 

Fixes #3710, #3480

nicojs commented:
There is a complication with this approach. It might result in this error: 

```
Cannot read properties of undefined (reading 'testEnvironmentOptions') TypeError: Cannot read properties of undefined (reading 'testEnvironmentOptions')
```

For both running regular unit tests as running stryker-js.

This happens when a user installed `jest@<=27`. See https://stackoverflow.com/questions/72078160/jest-testenvironmentoptions-cannot-be-read.

This happens because `jest-config@>29.3` is now a direct dependency of `@stryker-mutator/jest-runner`. It has `jest-environment-node` and `jest-environment-jsdom` listed as transient dependencies. This results in roughly this tree:

```
.
└── node_modules
    ├── @stryker-mutator
    │   └── jest-runner
    ├── jest@27
    │   └── node_modules
    │       └── jest-environment-node@27
    └── jest-environment-node@29.3
```

(or `jest-environment-jsdom` equivalent)

The way to solve this is actually not that big of a deal: install `jest-environment-node@27` yourself (or `jsdom` equivalent). This results in the following tree:

```
.
└── node_modules
    ├── @stryker-mutator
    │   └── jest-runner
    │       └── node_modules
    │           └── jest-environment-node@29
    ├── jest
    │   └── node_modules
    └── jest-environment-node@27
```

I still like the solution proposed in this PR. 

An alternative might be to use the preinstalled version of `jest-config` when `jest@>29.3` is detected. The downsides of that are:
- ⛔ reading `jest.config.ts` and other formats won't be supported for `jest@<29.3`.
- ⛔ a more significant maintenance burden on us.

@jimCresswell @bmordue @chaouiyounes11 what would you guys prefer?

@SimenB feel free to pitch in as well 😅

jimCresswell commented:
Thanks @nicojs !

> The way to solve this is actually not that big of a deal: install `jest-environment-node@27` yourself (or `jsdom` equivalent).

I prefer this solution as well, it is more transparent to the end user, and also gives them the opportunity to solve the problem by updating their dependencies.

I think the only caveat is that it needs to be well documented in a way that is easy to find with a quick search.

SimenB commented:
I'd try to load `jest-config` via the jest installation (meaning "reading `jest.config.ts` and other formats won't be supported for `jest@<29.3`" which is fine I think? Just document it).

```js
const {createRequire} = require('module');
const requireFromJest = createRequire(require.resolve('jest'));
const requireFromJestCli = createRequire(requireFromJest.resolve('jest-cli'));
const jestConfigFromJest = requireFromJestCli('jest-config');
// or  requireFromJestCli('jest-config/package.json'); if you wanna look at e.g. the exact version
```

---

We've talked about splitting out defaults from `jest-config` before, so that it can be used without also loading the envs etc.. Never got around to doing it, tho.

nicojs commented:
Thanks, @SimonB, and @jimCresswell for the advice.

I'm tempted to go with @SimonB's approach, simply because it would save a lot of headaches for existing users. One big downside is that `create-react-app` users don't have the luxury of choosing when to update jest... 😑

SimenB commented:
RN and Metro recently upgraded to Jest 29, so CRA might be able to as well 🙂 (assuming the holdup is Meta and not CRA itself)

nicojs commented:
I've implemented @SimenB 's suggestion. I also added custom support for exporting an `async` function from jest config, even when using older versions (implemented that while I was at it ;)).

jimCresswell commented:
:boom:

nicojs commented:
Released in v6.3 🎉
https://github.com/stryker-mutator/stryker-js/releases/tag/v6.3.0

nicojs commented:
Thanks for all help and feedback everyone! ♥
